### DISCUSSION TITLE
Support for TempId

### DISCUSSION THREAD

**Discussion by frankiesardo** [prisma/prisma#7577] [Q&A] [answered]
When inserting nested data, it's often useful to refer to entities you're inserting somewhere else in your data structure.

Consider this schema:

```
model Product {
  id             Int             @id @default(autoincrement())
  name           String
  options        Option[]
  configurations Configuration[]
}

model Option {
  id        Int           @id @default(autoincrement())
  name      String
  values    OptionValue[]
  Product   Product?      @relation(fields: [productId], references: [id])
  productId Int?
}

model OptionValue {
  id             Int             @id @default(autoincrement())
  name           String
  parent         Option          @relation(fields: [parentId], references: [id])
  parentId       Int
  configurations Configuration[]
}

model Configuration {
  id        Int           @id @default(autoincrement())
  product   Product       @relation(fields: [productId], references: [id])
  productId Int
  options   OptionValue[]
}
```

We want to be able to refer to the id of some of the option values we're inserting to create configurations linked to them.

```js
  await db.product.create({
    data: {
      name: "Product 1",
      options: {
        create: [
          {
            name: "Colour",
            values: {
              create: [
                { name: "Red" },
                { name: "Blue" },
              ],
            },
          },
          {
            name: "Size",
            values: {
              create: [
                { name: "Small" },
                { name: "Medium" },
              ],
            },
          },
        ],
      },
      configurations: {
        create: [
          { options: { connect: [{ id: ?red }, { id: ?small }] } },
          { options: { connect: [{ id: ?red }, { id: ?medium }] } },
          { options: { connect: [{ id: ?blue }, { id: ?small }] } },
          { options: { connect: [{ id: ?blue }, { id: ?medium }] } },
        ],
      },
    },
  })
```

At the moment there's no way to refer to red, blue, small, medium ids if you want to insert the configurations along with the rest of the product info.

Something like Datomic has a nice solution to this problem: using TempIds https://docs.datomic.com/on-prem/transactions/transactions.html#temporary-ids 
You can give a special value to the id of the entity you're transacting and refer to that in your nested data structure. When prisma is inserting the record and finds a tempid, it will swap it for the actual id of the inserted record. For example:

```js
  await db.product.create({
    data: {
      name: "Product 1",
      options: {
        create: [
          {
            name: "Colour",
            values: {
              create: [
                { id: -10, name: "Red" },
                { id: -11, name: "Blue" },
              ],
            },
          },
          {
            name: "Size",
            values: {
              create: [
                { id: -20, name: "Small" },
                { id: -21, name: "Medium" },
              ],
            },
          },
        ],
      },
      configurations: {
        create: [
          { options: { connect: [{ id: -10 }, { id: -20 }] } },
          { options: { connect: [{ id: -10 }, { id: -21 }] } },
          { options: { connect: [{ id: -11 }, { id: -20 }] } },
          { options: { connect: [{ id: -11 }, { id: -21 }] } },
        ],
      },
    },
  })
```

This works right now in Prisma but it will actually insert negative ids. It would be nice if there was a special convention e.g. if the id is autogenerated and you find a negative id instead, keep a tempId map and replace connected ids as needed, like so:

```js
const tempIds = { -10: 2345, -11: 2346, -20: 2347, -21: 2348 }
```

If a convention like "Negative id means tempid" is too much magic, the type system could be extended to support an explicit tempid on top of whatever the model id is: e.g. a number or a string:

```js
type Tempid = { tempid: string }
type OptionId = number | Tempid
```

The advantage of this solution is that you can give explicit temporary names to the entities you're transacting and refer to them in a clearer way:

```js
  await db.product.create({
    data: {
      name: "Product 1",
      options: {
        create: [
          {
            name: "Colour",
            values: {
              create: [
                { id: { tempid: "red" }, name: "Red" },
                { id: { tempid: "blue" }, name: "Blue" },
              ],
            },
          },
          {
            name: "Size",
            values: {
              create: [
                { id: { tempid: "small" }, name: "Small" },
                { id: { tempid: "medium" }, name: "Medium" },
              ],
            },
          },
        ],
      },
      configurations: {
        create: [
          { options: { connect: [{ id: { tempid: "red" } }, { id: { tempid: "small" }}] } },
          { options: { connect: [{ id: { tempid: "red" } }, { id: { tempid: "medium" }}] } },
          { options: { connect: [{ id: { tempid: "blue" } }, { id: { tempid: "small" }}] } },
          { options: { connect: [{ id: { tempid: "blue" }}, { id: { tempid: "medium" }}] } },
        ],
      },
    },
  })
```

I think this would be a great feature to add, particularly because Prisma already works so well with nested / graph like data structures and needs a way for these entities to reference each other.

Alternatively: if there's a way to accomplish this right now that I've missed (e.g. via middleware), please let me know

**Comment by ryands17** [comment, marked as answer]
@frankiesardo 👋 

I would suggest creating this as a feature request [here](https://github.com/prisma/prisma/issues/new?assignees=&labels=&template=feature_request.md&title=) 🙂

**Reply by frankiesardo** [reply]
Sure thing 👍

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
